### PR TITLE
+indexes for AgentDataSourceConfiguration and AgentConfiguration

### PR DESCRIPTION
**Pull request by lasryaric**
Adding indexes for two queries:
- [Query 1](https://github.com/dust-tt/dust/blob/2a28c677e3159bf6132d5f43c7a41cfa9244cf79/front/lib/api/assistant/configuration.ts#L92)
- [Query 2](https://github.com/dust-tt/dust/blob/2a28c677e3159bf6132d5f43c7a41cfa9244cf79/front/lib/api/assistant/configuration.ts#L219)

Query 1 has been the number 1 ressource consuming query in the front database for a little while now (as reported on [GCP query insight](https://console.cloud.google.com/sql/instances/dust-api-db/insights;database=dust-front;duration=PT1H;query_hash=2201675645681319626;sort_by=TOTAL_EXEC_TIME?project=or1g1n-186209)), and number 2 showed up as a pretty slow query (200 ms) on datadog.

# Deployment:

- Deploy to front
- Run initdb from front.

### CODE REVIEW

**Comment by lasryaric** [comment]
> Awesome! How much do we gain with the indexes? Have you made some local tests?

@PopDaph no but how do you test that locally? Just generate a lot of fake data?

The runtime complexity of these queries right now is not acceptable in any case, and they show up as top consuming ressources and they are not properly indexed, so spending time measuring it feels like a waste of time.

I did measure their cost in production without indexes though, and they are high.

**Comment by lasryaric** [comment]
I don't have the numbers easily accessible but if you need them I can work on getting them again.

**Comment by lasryaric** [comment]
[Query #1](https://metabase.dust.tt/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJleHBsYWluIGFuYWx5emVcblNFTEVDVCBcImFnZW50X2RhdGFfc291cmNlX2NvbmZpZ3VyYXRpb25cIi5cImlkXCIsXG4gICAgICAgXCJhZ2VudF9kYXRhX3NvdXJjZV9jb25maWd1cmF0aW9uXCIuXCJjcmVhdGVkQXRcIixcbiAgICAgICBcImFnZW50X2RhdGFfc291cmNlX2NvbmZpZ3VyYXRpb25cIi5cInVwZGF0ZWRBdFwiLFxuICAgICAgIFwiYWdlbnRfZGF0YV9zb3VyY2VfY29uZmlndXJhdGlvblwiLlwidGFnc0luXCIsXG4gICAgICAgXCJhZ2VudF9kYXRhX3NvdXJjZV9jb25maWd1cmF0aW9uXCIuXCJ0YWdzTm90SW5cIixcbiAgICAgICBcImFnZW50X2RhdGFfc291cmNlX2NvbmZpZ3VyYXRpb25cIi5cInBhcmVudHNJblwiLFxuICAgICAgIFwiYWdlbnRfZGF0YV9zb3VyY2VfY29uZmlndXJhdGlvblwiLlwicGFyZW50c05vdEluXCIsXG4gICAgICAgXCJhZ2VudF9kYXRhX3NvdXJjZV9jb25maWd1cmF0aW9uXCIuXCJyZXRyaWV2YWxDb25maWd1cmF0aW9uSWRcIixcbiAgICAgICBcImFnZW50X2RhdGFfc291cmNlX2NvbmZpZ3VyYXRpb25cIi5cImRhdGFTb3VyY2VJZFwiLFxuICAgICAgIFwiZGF0YVNvdXJjZVwiLlwiaWRcIiBBUyBcImRhdGFTb3VyY2UuaWRcIixcbiAgICAgICBcImRhdGFTb3VyY2VcIi5cImNyZWF0ZWRBdFwiIEFTIFwiZGF0YVNvdXJjZS5jcmVhdGVkQXRcIixcbiAgICAgICBcImRhdGFTb3VyY2VcIi5cInVwZGF0ZWRBdFwiIEFTIFwiZGF0YVNvdXJjZS51cGRhdGVkQXRcIixcbiAgICAgICBcImRhdGFTb3VyY2VcIi5cIm5hbWVcIiBBUyBcImRhdGFTb3VyY2UubmFtZVwiLFxuICAgICAgIFwiZGF0YVNvdXJjZVwiLlwiZGVzY3JpcHRpb25cIiBBUyBcImRhdGFTb3VyY2UuZGVzY3JpcHRpb25cIixcbiAgICAgICBcImRhdGFTb3VyY2VcIi5cInZpc2liaWxpdHlcIiBBUyBcImRhdGFTb3VyY2UudmlzaWJpbGl0eVwiLFxuICAgICAgIFwiZGF0YVNvdXJjZVwiLlwiYXNzaXN0YW50RGVmYXVsdFNlbGVjdGVkXCIgQVMgXCJkYXRhU291cmNlLmFzc2lzdGFudERlZmF1bHRTZWxlY3RlZFwiLFxuICAgICAgIFwiZGF0YVNvdXJjZVwiLlwiZHVzdEFQSVByb2plY3RJZFwiIEFTIFwiZFxuYXRhU291cmNlLmR1c3RBUElQcm9qZWN0SWRcIixcbiAgICAgICBcImRhdGFTb3VyY2VcIi5cImNvbm5lY3RvcklkXCIgQVMgXCJkYXRhU291cmNlLmNvbm5lY3RvcklkXCIsXG4gICAgICAgXCJkYXRhU291cmNlXCIuXCJjb25uZWN0b3JQcm92aWRlclwiIEFTIFwiZGF0YVNvdXJjZS5jb25uZWN0b3JQcm92aWRlclwiLFxuICAgICAgIFwiZGF0YVNvdXJjZVwiLlwid29ya3NwYWNlSWRcIiBBUyBcImRhdGFTb3VyY2Uud29ya3NwYWNlSWRcIixcbiAgICAgICBcImRhdGFTb3VyY2UtPndvcmtzcGFjZVwiLlwiaWRcIiBBUyBcImRhdGFTb3VyY2Uud29ya3NwYWNlLmlkXCIsXG4gICAgICAgXCJkYXRhU291cmNlLT53b3Jrc3BhY2VcIi5cImNyZWF0ZWRBdFwiIEFTIFwiZGF0YVNvdXJjZS53b3Jrc3BhY2UuY3JlYXRlZEF0XCIsXG4gICAgICAgXCJkYXRhU291cmNlLT53b3Jrc3BhY2VcIi5cInVwZGF0ZWRBdFwiIEFTIFwiZGF0YVNvdXJjZS53b3Jrc3BhY2UudXBkYXRlZEF0XCIsXG4gICAgICAgXCJkYXRhU291cmNlLT53b3Jrc3BhY2VcIi5cInVwZ3JhZGVkQXRcIiBBUyBcImRhdGFTb3VyY2Uud29ya3NwYWNlLnVwZ3JhZGVkQXRcIixcbiAgICAgICBcImRhdGFTb3VyY2UtPndvcmtzcGFjZVwiLlwic0lkXCIgQVMgXCJkYXRhU291cmNlLndvcmtzcGFjZS5zSWRcIixcbiAgICAgICBcImRhdGFTb3VyY2UtPndvcmtzcGFjZVwiLlwibmFtZVwiIEFTIFwiZGF0YVNvdXJjZS53b3Jrc3BhY2UubmFtZVwiLFxuICAgICAgIFwiZGF0YVNvdXJjZS0-d29ya3NwYWNlXCIuXCJkZXNjcmlwdGlvblwiIEFTIFwiZGF0YVNvdXJjZS53b3Jrc3BhY2UuZGVzY3JpcHRpb25cIixcbiAgICAgICBcImRhdGFTb3VyY2UtPndvcmtzcGFjZVwiLlwiYWxsb3dlZERvbWFpblwiIEFTIFwiZGF0YVNvdXJjZS53b3Jrc3BhY2UuYWxsb3dlZERvbWFpblwiLFxuICAgICAgIFwiZGF0YVNvdXJjZS0-d29ya3NwYWNlXCIuXCJzZWdtZW50YXRpb25cIiBBUyBcImRhdGFTb3VyY2Uud29ya3NwYWNlLnNlZ21lbnRhdGlvblwiXG5GUk9NIFwiYWdlbnRfZGF0YV9zb3VyY2VfY29uZmlndXJhdGlvbnNcIiBBUyBcImFnZW50X2RhdGFfc291cmNlX2NvbmZpZ3VyYXRpb25cIlxuTEVGVCBPVVRFUiBKT0lOIFwiZGF0YV9zb3VyY2VzXCIgQVMgXCJkYXRhU291cmNlXCIgT04gXCJhZ2VudF9kYXRhX3NvdXJjZV9jb25maWd1cmF0aW9uXCIuXCJkYXRhU291cmNlSWRcIiA9IFwiZGF0YVNvdXJjZVwiLlwiaWRcIlxuTEVGVCBPVVRFUiBKT0lOIFwid29ya3NwYWNlc1wiIEFTIFwiZGF0YVNvdXJjZS0-d29ya3NwYWNlXCIgT04gXCJkYXRhU291cmNlXCIuXCJ3b3Jrc3BhY2VJZFwiID0gXCJkYXRhU291cmNlLT53b3Jrc3BhY2VcIi5cImlkXCJcbldIRVJFIFwiYWdlbnRfZGF0YV9zb3VyY2VfY29uZmlndXJhdGlvblwiLlwicmV0cmlldmFsQ29uZmlndXJhdGlvbklkXCIgPSAxOyIsInRlbXBsYXRlLXRhZ3MiOnt9fSwiZGF0YWJhc2UiOjR9LCJkaXNwbGF5IjoidGFibGUiLCJwYXJhbWV0ZXJzIjpbXSwidmlzdWFsaXphdGlvbl9zZXR0aW5ncyI6e319) does a full scan table.

**Comment by lasryaric** [comment]
I agree that until we test we don't know 100% the result, but I just don't know how to correctly test these in production.
The query planner makes very different decisions depending on the amount of data and the variety of the data, so faking it with random data with one or two accounts / workspaces / data sources is likely to trigger different results from production runtime, hence my skepticism about testing locally.

**Comment by PopDaph** [comment]
> I don't have the numbers easily accessible but if you need them I can work on getting them again.

I don't _need_ anything I was just wondering if we were able to measure what gain we expect from those indexes! Your answer is perfectly fine to me 👍🏻

**Comment by lasryaric** [comment]
@PopDaph I found a problem in my index for query #1 while trying to demonstrate that it was correct. Thanks @PopDaph 🙏

**Comment by lasryaric** [comment]
And for the record:
Query #1 goes from O(n) to O(log(n)) with n being the number of elements in the table `agent_data_source_configurations`
Query #2 moves from O(n) to O(log(n)) with n being the number of entires for a given workspace.

File: front/lib/models/assistant/agent.ts [hunk 1]
```diff
@@ -174,6 +174,7 @@ AgentConfiguration.init(
     indexes: [
       { fields: ["workspaceId"] },
       { fields: ["workspaceId", "name"] },
+      { fields: ["workspaceId", "status", "name"] },
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply front/lib/models/assistant/agent.ts:177]: We filter per workspace, status and sort by name, which is a perfect use case for a composit index that [workspaceId, status, name].

[the remark this answers is not in the document]
Comment (PopDaph) [reply front/lib/models/assistant/agent.ts:177]: Indeed we sort always, make sense! Thanks 🙏🏻

**Comment by lasryaric** [comment]
Yay! Switched to using index vs full table scan for query #1, with an actual complexity of 16 vs 138 (unit is "unknown", but who cares!)

**Pull request merged**